### DISCUSSION TITLE
Softkeyboard shrinks the screen

### DISCUSSION THREAD

**Discussion by abibik1** [airsdk/Adobe-Runtime-Support#2356] [General]
Hi,

I need help, please.
I am building app for Android, using AIR SDK 50.0.1.3 targeting android:targetSdkVersion="29".
On Android 13 once an input controls receives focus, a soft keyboard is shown. But the soft keyboard shrinks the the view distorting all controls. Sometimes the view shrinks for a split second, sometimes - for a couple of seconds.
Look at this video: https://www.youtube.com/shorts/dPcpyiiq0nU

I can't give it to a customer as is. I need to fix that shrinkage. When I was building using AIR 33, the screen will get shrinked and will stay shrinked. With AIR 50 it is a bit better as the screen recovers, but still that flickering can't be shown to a customer.
It happens only on Android 13

This parameter in manifest is getting completely ignored.
<softKeyboardBehavior>pan</softKeyboardBehavior>

I can't specify android:windowSoftInputMode="adjustPan" (https://developer.android.com/guide/topics/manifest/activity-element) in android manifest as this setting applies to the main activity which is generated by AIR and I can't add any parameters there.

Please, advise.

Andrew.

**Comment by abibik1** [comment]
Anyone?
I upacked .apk and foun there this:
![image](https://user-images.githubusercontent.com/49885119/208765258-50c10611-54a5-4eac-82e3-eb9630136f41.png)
Looks like a bug to me...

**Reply by ajwfrost** [reply]
Looking at our templates, that value should be `adjustResize|stateHidden` which is what we want it to be:
https://developer.android.com/guide/topics/manifest/activity-element#wsoft

The UI flickering/updates are presumably then caused by the resize handlers and the fact we need to adjust the AIR stage size and request the application to redraw itself..

If you want to override this parameter so that it just does the pan rather than the resize, you could try using the `android-studio` target build type (rather than `apk` or `aab`) and then ADT will generate you an Android Studio project. You can then open that up, edit the manifest file manually, and build the apk file using Android Studio..?

thanks

**Reply by abibik1** [reply]
Hi!

Thanks for your reply... I thought the pan/resize keyboard behavior is controlled via Application descriptor:
![image](https://user-images.githubusercontent.com/49885119/208941268-8006c0da-5e82-4dbb-a81f-e817ef70aab3.png)
Is this setting being deprecated?

Thanks for the idea to build it as android-studio project! I had no idea about this option. Any hint how to do it in IntelliJ? My project have 3 sub-projects, more than 1000 source files. I don't really want to build it manually...

**Reply by ajwfrost** [reply]
Wow I had no idea about that setting!! Now I see what you mean by the comment in the original post... Let me check further on this..

To get from IntelliJ to Android Studio builds, the only thing I can suggest is to enable logging in the ADT tool and then look to see the command line that IntelliJ has done; then run that from the appropriate location in a command prompt and change the "apk" target to "android-studio".

thanks

**Reply by ajwfrost** [reply]
So it seems there may be something else going on .. the 'pan' behaviour is used to scroll an area of interest into view when the soft keyboard has been launched, but I think there is going to be separate logic happening where the activity is being resized. So it's likely that there's both a resize of the view area - resulting in (initially) squished-looking display and then the application layout code is presumably kicking in to sort it out - plus then there's the pan of the content to ensure the chosen text field is visible (which in the video is probably a no-op as it's already shown..)

I'm just wondering whether that app descriptor setting should then also be used to change the behaviour of the activity... we can check a bit further on this to see whether it helps in this scenario.

thanks

**Comment by abibik1** [comment]
Hi,
In my experience the <softKeyboardBehaviour> parameter in app descriptor does not work at all since at least AIR SDK 33.

**Comment by phitrti** [comment]
Andrew,
I am running into a similar issue where, on Android, the soft keyboard display smushes the UI and the actual input text for the entry field doesn't display in the smushed entry field but rather just below it. softKeyboardBehavior is not set so we should be seeing, presumably, whatever the default behavior is for Android. I'll continue to dig but if you have any info that would help, much appreciated. Current SDK is AIR50.2.3.2. On the radar to upgrade, just haven't made it happen yet. 
![image-2024-02-22-15-12-48-169](https://github.com/airsdk/Adobe-Runtime-Support/assets/21318472/f3cf31cc-e549-4c04-a944-7dbd2e83c305)

**Reply by phitrti** [reply]
@ajwfrost  Just saw your post about Text. I didn't see mention about spark.components.TextInput. Granted, that is for input but StageText is also input. So..... i looked for the spark components in https://airsdk.dev/reference and didn't see them. Are spark components no longer supported? Egg on my face for missing that one if that is true. We are using spark.components.TextInput for our input field and many of our mobile UX elements use spark.components.

**Reply by ajwfrost** [reply]
Hi

We're still working on this soft keyboard behaviour, primarily aiming to get the event values working from a 'raw' AIR perspective... which hopefully will mean it then also looks good for Flex, but there is a possibility that they'd already been trying to work around the Flash/AIR issues so may mean changes are needed there.

For Flex - including the documentation for spark and mx components etc - this was all handed off to Apache quite a few years ago. Adobe are still maintaining the documentation for it on their website but I don't know for how much longer. As we don't maintain Flex, we don't publish the documentation for it, but it is available from the Apache website:
https://flex.apache.org/asdoc/

As to whether the Flex components are supported .. well they are, loosely I would say, but they're also fully open source, so anyone can review/update/submit changes/fork and build, etc...

thanks

**Reply by abibik1** [reply]
To solve our problem we use "magic" combination of settings.

The manifest:
<renderMode>gpu</renderMode>
<softKeyboardBehavior>pan</softKeyboardBehavior>

Also we use this skin specified in CSS:
s|TextInput { skinClass: ClassReference("spark.skins.mobile.TextInputSkin");}

All these 3 parameters combined render acceptable result.
For your scenario use appropriate skin from spark.skins.mobile.* 

It MAY work.
Good luck!
